homeassistant/components/google_assistant_sdk/__init__.py

```diff
@@ -90,3 +96,64 @@ async def send_text_command(call: ServiceCall) -> None:
         send_text_command,
         schema=SERVICE_SEND_TEXT_COMMAND_SCHEMA,
     )
+
+
+async def update_listener(hass, entry):
+    """Handle options update."""
+    if entry.options.get(CONF_ENABLE_CONVERSATION_AGENT, False):
+        agent = GoogleAssistantConversationAgent(hass, entry)
+        conversation.async_set_agent(hass, agent)
+    else:
+        conversation.async_set_agent(hass, None)
+
+
+class GoogleAssistantConversationAgent(conversation.AbstractConversationAgent):
+    """Google Assistant SDK conversation agent."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the agent."""
+        self.hass = hass
+        self.entry = entry
+        self.assistant: TextAssistant | None = None
+        self.session: OAuth2Session | None = None
+
+    @property
+    def attribution(self):
+        """Return the attribution."""
+        return {
+            "name": "Powered by Google Assistant SDK",
+            "url": "https://www.home-assistant.io/integrations/google_assistant_sdk/",
+        }
+
+    async def async_process(
+        self,
+        text: str,
+        context: Context,
+        conversation_id: str | None = None,
+        language: str | None = None,
+    ) -> conversation.ConversationResult | None:
+        """Process a sentence."""
+        if self.session:
+            session = self.session
+        else:
+            session = self.hass.data[DOMAIN].get(self.entry.entry_id)
+            self.session = session
+        if not session.valid_token:
+            await session.async_ensure_token_valid()
+            self.assistant = None
+        if not self.assistant:
+            credentials = Credentials(session.token[CONF_ACCESS_TOKEN])
+            language_code = self.entry.options.get(
+                CONF_LANGUAGE_CODE, default_language_code(self.hass)
+            )
+            self.assistant = TextAssistant(credentials, language_code)
+
+        resp = self.assistant.assist(text)
+        text_response = resp[0]
+
+        language = language or self.hass.config.language
+        intent_response = intent.IntentResponse(language=language)
+        intent_response.async_set_speech(text_response)
+        return conversation.ConversationResult(
+            response=intent_response, conversation_id=conversation_id
```

Comment:
The way `conversation_id` is used that when a conversation is initiated, it is set to None.

Then in your response you return a conversation ID such that follow-up responses work.

Now in this case it doesn't seem like you're passing conversation_id so you should set `conversation_id=None` here, as it's not used.
```suggestion
            response=intent_response, conversation_id=None
```

Now in your screenshot I did notice that it's allowing follow-up commands, so I am not sure how that works then 🤷

Reply:
Both DefaultAgent and AlmondAgent (the only other two agents) do exactly the same thing. conversation_id starts with None and is never modified.

Anyway for Google Assistant we cannot really have many concurrent conversations, see https://github.com/tronikos/gassist_text/blob/main/src/google/assistant/embedded/v1alpha2/embedded_assistant.proto#L314-L318

Reply:
Yeah neither can do conversations.